spacepy/pybats/__init__.py

```diff
@@ -877,49 +1093,249 @@ class IdlFile(PbData):
+                [self.attrs['time_range'][0]+tdelt(seconds=int(x)) for x in runtimes])
+        else:
+            self.attrs['times'] = np.array(nframe*[None])
+
+        # Ensure all ranges are two-element arrays and update using
+        # information we gathered from the header.
+        if self.attrs['iter_range'] == [None]:
+            self.attrs['iter_range'] = [iters[0], iters[-1]]
+        if self.attrs['runtime_range'] == [None]:
+            self.attrs['runtime_range'] = [runtimes[0], runtimes[-1]]
+
+        # Stash the offset of frames as a private attribute:
+        self._offsets = np.array(offset)
+    
+    def _scan_asc_frames(self):
+        '''
+        Open the ascii-formatted file associated with *self* and scan all
+        headers to count the number of epoch frames and details of each.
+        Results are stored within *self*.
 
+        This is a placeholder only until ascii-formatted .outs files are
+        fully supported.
+        '''
+
+        # Only use top-level frame as of now.
+        self._offsets = np.array([0])
+        self.attrs['nframe']   = 1
+        self.attrs['iters']    = [0,0]
+        self.attrs['runtimes'] = [0,0]
+        self.attrs['times']    = [0,0]
+    
+    def switch_frame(self, iframe):
+        '''
+        For files that have more than one data frame (i.e., `*.outs` files),
+        load data from the *iframe*-th frame into the object replacing what is
+        currently loaded.
+        '''
+
+        # Ensure that given frame exists within object:
+        if iframe < -self.attrs['nframe'] or iframe >= self.attrs['nframe']:
+            raise IndexError("iframe {} is outside range of [0,{})".format(
+                iframe, self.attrs['nframe']))
+        
+        self.read(iframe)
+
+        # Update information about the current frame:
+        self.attrs['iframe'] = self.attrs['nframe'] + iframe if iframe < 0 else iframe
+        self.attrs['iter']   = self.attrs['iters'][iframe]
+        self.attrs['runtime']= self.attrs['runtimes'][iframe]
+        self.attrs['time']   = self.attrs['times'][iframe]
+
+        # Update string time:
+        time = self.attrs['runtime']  # Convenience variable.
+        self.attrs['strtime'] = "{:04.0f}h{:02.0f}m{:06.3f}s".format(
+            np.floor(time//3600), np.floor(time % 3600//60.0), time % 60.0)
+
+        # Re-do calculations if any have been done.
+        if hasattr(self, '_calcs'):
+            # Get dictionary of name-method pairs for called calculations:
+            calcs = self._calcs
+            # Reset dict of called calculations:
+            self._calcs = {}
+            # Call all calculation methods:
+            for name in calcs:
+                args, kwargs = calcs[name]
+                getattr(self, name)(*args, **kwargs)
+                
     def __repr__(self):
         return 'SWMF IDL-Binary file "%s"' % (self.attrs['file'])
-    
-    def read(self, header, keep_case):
+
+    def read(self, iframe=0):
```

Comment:
While I'm thinking about feature requests to make this easier to use - it might be useful to be able to specify a time and get the nearest frame (or just have a convenience function for frame index lookup).

Reply:
Speaking from experience, the latter would be most useful.  Will take a bit of work/thought.  It's also a pandora's box to interpolating to an arbitrary time between two frames -- super useful and something I've considered, but for a later date.